### PR TITLE
persistence test

### LINKED ISSUE #6502: Intg test: Persistence Test
We need a test that ensures that a node can restart from a persisted state without needing to bootstrap back to the network (for low downtimes of nodes, anyway).

Long term, there are lots of edge cases we want to test with persistence, such as killing the node in different ways and at different block heights to exhibit different code paths, but even a simple test will suffice for now.

To test this, we will want to stop and start a node that is synced to a running chain (using the puppeteer start/stop, not restarting the kubernetes pod), and monitor that node to ensure that it syncs to the network within an expected amount of time. Additionally, the test should look for events in the logs to ensure that:

1) the node loaded the persistent frontier from disk
2) the node did not throw the persistent frontier away
3) the node synced without entering bootstrap

All the logs for these exist, but they will need to be converted into structured logs for the integration test framework.

### PR DESCRIPTION
**Pull request by dkijania**
Explain your changes:

New verification for medium-bootstrap regarding bootstrap from already existing storage. For this purpose I've added new structured log and new assertion.

Explain how you tested your changes:

Run medium-bootstrap test with new code

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

* Closes #6502

### CODE REVIEW

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

File: src/lib/integration_test_cloud_engine/graphql_polling_log_engine.ml [hunk 1]
```diff
@@ -33,13 +33,9 @@ let event_reader { event_reader; _ } = event_reader
...
 let parse_event_from_log_entry ~logger log_entry =
   let open Or_error.Let_syntax in
-  let open Json_parsing in
   Or_error.try_with_join (fun () ->
-      let payload = Yojson.Safe.from_string log_entry in
       let%map event =
-        let%bind msg =
-          parse (parser_from_of_yojson Logger.Message.of_yojson) payload
-        in
+        let%bind msg = Util.log_entry_from_string log_entry in
```
Comment (QuiteStochastic) [inline src/lib/integration_test_cloud_engine/graphql_polling_log_engine.ml:38]: you can make this into a helper function if you like, but don't put it in util.ml. put the helper function in this same file, graphql_polling_log_engine.ml.

util.ml is more for functions which are going to be shared across files, this log parsing stuff is pretty specific to the log engines and we only use this one log engine atm

Comment (dkijania) [reply src/lib/integration_test_cloud_engine/graphql_polling_log_engine.ml:38]: @QuiteStochastic I'm also using this method in test_common module in test_executive app. Will it be available to that module after i move this method to graphql_polling_log_engine.ml (without adding an interface method to intf in integration_test_lib )  ?

**Comment by QuiteStochastic** [comment]
Hi @dkijania

From a design point of view, we should not be calling the logging engine from the app/test_executive directory. The log parsing is a low-ish level implementation detail, in the future we may in fact replace the graphql logging engine and if we did, then this would break this entire test. In other words, the log engine is a part of the infrastructure engine, but the test_executive (including individual tests, as well as test_common.ml) should be only interacting with the DSL.

I realize that it's a tad mixed as it is because graphql in general is a part of the DSL and it's ok for tests to use graphql normally, but even though the log engine uses graphql, that specific bit should stay within the infrastructure layer and not the DSL layer, at least conceptually.

I would design this by hooking up a new event type, or multiple event types, surrounding the persisted frontier being loaded. there needs to be a new event router which listens for these events, and modifies the network_state when these events happen. you can add a new data structure into network_state which tracks how many times the persisted frontier has been loaded (or something like that, i'm not entirely sure).

This is more along the lines of how lucy is meant to be used. Logs are fetched by the logs engine, they are parsed into events, event routers listen for events and modify the network state in response. tests should not be directly parsing logs, checking for log lines.

As such at the current moment I would not merge this.

Also I do believe that the persisted frontier is something that also exists in compatible yes? if so then this should be developed against compatible, and when that's done then merged against berk and dev.

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

File: src/lib/integration_test_lib/wait_condition.mli [hunk 2]
```diff
@@ -31,6 +31,7 @@ module Make
     | Block_height_growth
     | Zkapp_to_be_included_in_frontier
     | Persisted_frontier_loaded
+    | Node_to_bootstrap
```
Comment (QuiteStochastic) [inline src/lib/integration_test_lib/wait_condition.mli:34]: can you just add a comment explaining what "bootstrap" means in this context, and also explain how this is different from "Node_to_initialize".

File: src/lib/integration_test_lib/network_state.ml [hunk 3]
```diff
@@ -28,6 +28,10 @@ module Make
     ; global_slot : int
     ; snarked_ledgers_generated : int
     ; blocks_generated : int
+    ; persisted_frontier_loaded : int
+    ; persisted_frontier_fresh_boot : int
+    ; bootstrap_required : int
+    ; persisted_frontier_dropped : int
```
Comment (QuiteStochastic) [inline src/lib/integration_test_lib/network_state.ml:34]: this is a bit of a nitpick but all of these variable names sound like they should be booleans with true/false, like  "persisted_frontier_loaded", however instead they are ints.  can you rename them something better?  maybe "num_times_persisted_frontier_loaded" or something like that, if you can find something terser that'd be good.

File: src/lib/integration_test_lib/wait_condition.mli (hunk 2 above)
Comment (dkijania) [reply src/lib/integration_test_lib/wait_condition.mli]: Good point. It hard to differentiate currently as from technical side those events are basically for node initialization but are interested in different details of initialization. We have Persistent_frontier_loaded event which is a lowest level of node startup (in all initialization related events) then on slightly upper level we have 'Node_to_bootstrap' which is interested in the same event as `Persistent_frontier_loaded` but additionally want to learn about couple more information (if node dropped persistent frontier etc). Finally we have Node_to_initialize which doesn't care about those details and represent (i think) event which tells us about general node readiness.

IMHO, maybe if we rename `Node_to_bootstrap` which is too general to more specific like: `Transition_frontier_loaded` which is right name for current level of events, it will correctly inform anyone on which level of node bootstrap we are operating and what to expect. it will guide anyone that it is some more detailed than `Node_to_initialize` but no so much as  `Persisted_frontier_loaded`. what you think?

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

File: src/lib/integration_test_lib/wait_condition.mli (hunk 2 above)
Comment (QuiteStochastic) [reply src/lib/integration_test_lib/wait_condition.mli]: On the whole, I agree that renaming it `Transition_frontier_loaded` and `Persisted_frontier_loaded` is better, since it is more detailed, however there are some confusing wrinkles here.

the wait condition `Nodes_to_initialize` relies on the lucy event `Node_initialization`, which looks at the daemon structured event called `Starting_transition_frontier_controller`.  This also has to do with the transition frontier being loaded, although I assume this is a separate structured event?

I'm a little unclear as to how those structured events are different, although this is a daemon things and not a lucy thing per se.  Do you have any further information on this?

**Comment by dkijania** [comment]
!ci-build-me

File: src/lib/integration_test_lib/wait_condition.mli (hunk 2 above)
Comment (dkijania) [reply src/lib/integration_test_lib/wait_condition.mli]: Pasting @nholland94 response for future reference:
```
From how this code is written, Transition_frontier_loaded will only get triggered if the transition frontier was successfully loaded from persistence. I believe this will happen even before bootstrap, and then the bootstrapping process will determine if that loaded frontier is too out of date (in which case we would actually run bootstrap). The Starting_transition_frontier_controller will fire whenever we finish bootstrapping and begin participating in the network, regardless of whether we used persisted frontier or executed the full bootstrap routine.
```

**Comment by dkijania** [comment]
!ci-build-me

**Review by QuiteStochastic — APPROVED** [review]

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by dkijania** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
Looks great! 
Do you think (as a separate task) we could add this persistence section to payments and zkApp tests when they are waiting for ledger proof? I wouldn't want to increase the test time further but it'd be super useful to test persistence with transactions in blocks. 
It'd also be good to test persistence with updated snarked ledger but you'll probably want to trigger that `k` blocks after ledger proof is emitted.  It might be a good addition to the bootstrap with epoch ledger sync test.

**Comment by deepthiskumar** [comment]
!approved-for-mainnet

**Comment by deepthiskumar** [comment]
!ci-build-me
